### ISSUE TITLE
After updating it via AUR

### ISSUE THREAD

**Issue by niksingh710** [eylles/pywal16#48]
```
󰘧 walset
[I] image: Using image wallhaven-x67oxo.png.
[I] colors: Generating a colorscheme.
[I] colors: Using wal backend.
Traceback (most recent call last):
  File "/usr/bin/wal", line 33, in <module>
    sys.exit(load_entry_point('pywal16==3.5.3', 'console_scripts', 'wal')())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/pywal/__main__.py", line 244, in main
    parse_args(parser)
  File "/usr/lib/python3.12/site-packages/pywal/__main__.py", line 184, in parse_args
    colors_plain = colors.get(image_file, args.l, args.cols16, args.backend,
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/pywal/colors.py", line 213, in get
    colors = getattr(backend, "get")(img, light, cols16)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/pywal/backends/wal.py", line 77, in get
    colors = gen_colors(img)
             ^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/pywal/backends/wal.py", line 64, in gen_colors
    return [re.search("#.{6}", str(col)).group(0) for col in raw_colors[1:]]
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'group'
```

**Comment by niksingh710** [comment]
if i am using `haishoku` backend then there is no error but color generated is too bad

**Comment by CapybaraWarrior** [comment]
I am getting a similar error when trying to install with yay on artix

`> Generate a colorscheme. ... Generating a colorscheme.
Using wal backend.
ERROR
test_gen_colors_checksum (tests.test_colors.TestGenColors.test_gen_colors_checksum)
> Generate a colorscheme with the wallpaper's checksum ... Generating a colorscheme.
Using wal backend.
ERROR

======================================================================
ERROR: test_gen_colors (tests.test_colors.TestGenColors.test_gen_colors)
> Generate a colorscheme.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/theuser/.cache/yay/pywal-16-colors/src/pywal16/tests/test_colors.py", line 13, in test_gen_colors
    result = colors.get("tests/test_files/test.jpg")
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/theuser/.cache/yay/pywal-16-colors/src/pywal16/pywal/colors.py", line 213, in get
    colors = getattr(backend, "get")(img, light, cols16)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/theuser/.cache/yay/pywal-16-colors/src/pywal16/pywal/backends/wal.py", line 77, in get
    colors = gen_colors(img)
             ^^^^^^^^^^^^^^^
  File "/home/theuser/.cache/yay/pywal-16-colors/src/pywal16/pywal/backends/wal.py", line 64, in gen_colors
    return [re.search("#.{6}", str(col)).group(0) for col in raw_colors[1:]]
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'group'

======================================================================
ERROR: test_gen_colors_checksum (tests.test_colors.TestGenColors.test_gen_colors_checksum)
> Generate a colorscheme with the wallpaper's checksum
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/theuser/.cache/yay/pywal-16-colors/src/pywal16/tests/test_colors.py", line 38, in test_gen_colors_checksum
    result = colors.get("tests/test_files/test.jpg")
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/theuser/.cache/yay/pywal-16-colors/src/pywal16/pywal/colors.py", line 213, in get
    colors = getattr(backend, "get")(img, light, cols16)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/theuser/.cache/yay/pywal-16-colors/src/pywal16/pywal/backends/wal.py", line 77, in get
    colors = gen_colors(img)
             ^^^^^^^^^^^^^^^
  File "/home/theuser/.cache/yay/pywal-16-colors/src/pywal16/pywal/backends/wal.py", line 64, in gen_colors
    return [re.search("#.{6}", str(col)).group(0) for col in raw_colors[1:]]
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'group'

----------------------------------------------------------------------
Ran 31 tests in 0.086s

FAILED (errors=2)
Test failed: <unittest.runner.TextTestResult run=31 errors=2 failures=0>
error: Test failed: <unittest.runner.TextTestResult run=31 errors=2 failures=0>
==> ERROR: A failure occurred in check().
    Aborting...
 -> error making: pywal-16-colors-exit status 4
 -> Failed to install the following packages. Manual intervention is required:
pywal-16-colors - exit status 4`

**Comment by Foxap** [comment]
Same issue here , but I did not install it with the AUR , I used the pipx command to install pywal16 . It's exactly the same error  . I also had to copy the wal file in the sbin directory .

**Comment by niksingh710** [comment]
@eylles check this issue once

**Comment by eylles** [comment]
So weird, according to that the regular expression search `return [re.search("#.{6}", str(col)).group(0) for col in raw_colors[1:]]` is not getting any match for the group `#.{6}` which would be a # followed by any 6 chars.

What is the output of running `convert` or `magick convert` with the following options `convert <image> -resize 25% -colors 16 -unique-colors txt:-`

I know arch should be on imagemagick 7 but no idea if it does already ship with the new policy to not allow writing to std out.

To change the policy run: `sudo sed -i 's/\(<policy domain="path" rights="\)none\(" pattern="-"\/>\)/\1write\2/' /etc/ImageMagick-7/policy.xml`

If that does nothing i'll work on this tomorrow, but it won't be an easy matter since i'm on devuan testing and need to do upgrade packages next month.

**Comment by eylles** [comment]
Dumb question, does this also happen with dylan's pywal? be it 3.3.0 or pywal-git? That way at least i'd know if it was sonething i changed or is something that pywal has been dragging for years waiting to implode.

**Comment by CapybaraWarrior** [comment]
Just tried it and those 2 do not give an error when installing

**Comment by eylles** [comment]
@CapybaraWarrior do either of those give the runtime error with the traceback to:
```
return [re.search("#.{6}", str(col)).group(0) for col in raw_colors[1:]]
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'group'
```

**Comment by CapybaraWarrior** [comment]
You mean python-pywal and pywal-git? They install fine with yay and work without any error. Sorry, I should have said that they install and work without error instead of just that they install.

**Comment by eylles** [comment]
> You mean python-pywal and pywal-git? They install fine with yay and work without any error. Sorry, I should have said that they install and work without error instead of just that they install.

Just for completeness sake, you installed eother python-pywal or pywal-git from the aur, ran `wal -c` to erase current colorschemes and then ran `wal -i image` and the program generated a colorscheme without errors? Both should reply with 3.3.0 when you run `wal -v`

It is very important as there have been 7 changes to backends/wal.py since pywal's 3.3.0 to pywal16's 3.5.3, and since this is an arch specific (it seems, as no other distro has suffered from this yet) it helps to know at which commit does it begin to happen.

And no, i don't mean the aur install bug, that is something that should be reported to the aur maintainer, i mean the runtime traceback error as that is also reported by users instally the pypi package.

**Comment by CapybaraWarrior** [comment]
I tried it again with the options you mentioned and python-pywal gave no errors when I gave it an image and it worked, -v said 3.3.0 and then I ran -c and removed python-pywal and installed pywal-git. For python-git -v said 3.3.1 but it worked as well and there was this:
[I] colors: Using wal backend.
WARNING: The convert command is deprecated in IMv7, use "magick"

But otherwise no diffreence

**Comment by eylles** [comment]
> I tried it again with the options you mentioned and python-pywal gave no errors when I gave it an image and it worked, -v said 3.3.0 and then I ran -c and removed python-pywal and installed pywal-git. For python-git -v said 3.3.1 but it worked as well and there was this: [I] colors: Using wal backend. WARNING: The convert command is deprecated in IMv7, use "magick"
> 
> But otherwise no diffreence

Great, that DOES help a lot, just as a question, you can do clone the repo and do a `pipx install .` to install a different branch? Cuz i got an idea of where the runtime error may come from, but cannot push a fix that may not even work to master less alone to a release before testing.

For now while i prepare the branch anyone willing to help please install pipx from your package manager and make sure you uninstalled both pywal and pywal16.

**Comment by CapybaraWarrior** [comment]
So after testing some releases and installing them with pipx, all work and give no error when producing 9 colors but when I used --cols16 on version 3.5.3 I got this:

raceback (most recent call last):
  File "/home/theuser/.local/bin/wal", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/theuser/.local/share/pipx/venvs/pywal16/lib/python3.12/site-packages/pywal/__main__.py", line 244, in main
    parse_args(parser)
  File "/home/theuser/.local/share/pipx/venvs/pywal16/lib/python3.12/site-packages/pywal/__main__.py", line 184, in parse_args
    colors_plain = colors.get(image_file, args.l, args.cols16, args.backend,
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/theuser/.local/share/pipx/venvs/pywal16/lib/python3.12/site-packages/pywal/colors.py", line 213, in get
    colors = getattr(backend, "get")(img, light, cols16)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/theuser/.local/share/pipx/venvs/pywal16/lib/python3.12/site-packages/pywal/backends/wal.py", line 77, in get
    colors = gen_colors(img)
             ^^^^^^^^^^^^^^^
  File "/home/theuser/.local/share/pipx/venvs/pywal16/lib/python3.12/site-packages/pywal/backends/wal.py", line 64, in gen_colors
    return [re.search("#.{6}", str(col)).group(0) for col in raw_colors[1:]]
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'group'

**Comment by CapybaraWarrior** [comment]
Using version 3.5.2 works perfectly after installing with pipx

**Comment by eylles** [comment]
Thanks, now at least i know it was a change in 3.5.3 that caused this, give me some mins and i will have a branch to test.

**Comment by DerTyp7214** [comment]
Hotfix if anyone wants it: in `/usr/lib/python3.12/site-packages/pywal/backends/wal.py` remove `convert` arg from magick and it works (if you have imagemagick 7+)

**Comment by eylles** [comment]
okay, before merging anything to master i need to know that this branch fixes the issue https://github.com/eylles/pywal16/tree/arch-fix 

i think the correct command with pipx is `pipx install git+https://github.com/eylles/pywal16.git@arch-fix` to skip the whole clone, checkout branch and running `pipx install .`

**Comment by DerTyp7214** [comment]
Seems to work
![image](https://github.com/eylles/pywal16/assets/37804065/55f0f9e3-2e5f-412f-a7a4-2bd10a4ababb)

**Comment by eylles** [comment]
nice, to speak about the solution.

```py
def try_gen_in_range(img, magick_command):
    for i in range(0, 20, 1):
        raw_colors = imagemagick(16 + i, img, magick_command)

        if len(raw_colors) > 16:
            break

        if i == 19:
            logging.error("Imagemagick couldn't generate a suitable palette.")
            sys.exit(1)

        else:
            logging.warning("Imagemagick couldn't generate a palette.")
            logging.warning("Trying a larger palette size %s", 16 + i)
    return raw_colors


def gen_colors(img):
    """Format the output from imagemagick into a list
       of hex colors."""
    magick_command = has_im()

    raw_colors = try_gen_in_range(img, magick_command)

    try:
        out = [re.search("#.{6}", str(col)).group(0) for col in raw_colors[1:]]
    except AttributeError:
        if magick_command == ["magick", "convert"]:
            logging.warning("magick convert failed, using only magick")
            magick_command = ["magick"]
            raw_colors = try_gen_in_range(img, magick_command)
            out = [re.search("#.{6}",
                             str(col)).group(0) for col in raw_colors[1:]]

    return out

```

had to move the generation of colors in a for loop for continuous large palettes into it's own function.

used a try except to catch the attribute error from the regex to then try the magick command without the "convert" part.

the idea is this won't affect distros that do got imagemagick 7 but is an older version than what arch ships so it will just work on those.

**Comment by niksingh710** [comment]
Does aur have any `-git` version? If not then when this will be merged in tagged release?

**Comment by eylles** [comment]
well, i want to address not just this but also #31 and hopefully #49 as part of 3.5.4, if not i will only do #31 and take care of #49 for 3.6.0 which will entice the okthief backend from #47 and being able to separate the 16 color schemes in ones generated with the lighten or the darken method.

**Comment by eylles** [comment]
as everyone can see, i marked everything else as a milestone for 3.6.0 and released 3.5.4, please feel free to flood the aur maintainer for the latest release version.

if anyone wonders, yes constant email spam and flood is one of the reasons dylan's email is still the one publically associated with pywal16, but the main reason is i am lazy.
